platform/security/session-management-impl/src/main/java/org/codice/ddf/security/session/management/impl/SessionManagementServiceImpl.java

```diff
@@ -105,34 +76,52 @@ public Response getRenewal(@Context HttpServletRequest request) {
                   try {
                     doRenew(s, realmTokenMap.get(s), tokenHolder);
                   } catch (SecurityServiceException e) {
-                    response[0] = Response.serverError().build();
+                    securityServiceExceptionThrown[0] = true;
                     LOGGER.error("Failed to renew", e);
                   }
                 });
-        if (response[0] == null) {
-          response[0] =
-              Response.ok(
-                      new ByteArrayInputStream(
-                          Long.toString(getTimeLeft(tokenHolder)).getBytes(StandardCharsets.UTF_8)))
-                  .build();
+
+        if (securityServiceExceptionThrown[0]) {
+          return null;
         }
+
+        timeLeft = Long.toString(getTimeLeft(tokenHolder));
       }
     }
-    return response[0];
+    return timeLeft;
   }
 
-  @GET
-  @Path("/invalidate")
-  public Response getInvalidate(@Context HttpServletRequest request) {
-    StringBuffer requestURL = request.getRequestURL();
+  @Override
+  public URI getInvalidate(HttpServletRequest request) {
     String requestQueryString = request.getQueryString();
-    return Response.seeOther(
-            URI.create(
-                requestURL
-                    .substring(0, requestURL.indexOf(rootContext))
-                    .concat("/logout?noPrompt=true")
-                    .concat(requestQueryString != null ? "&" + requestQueryString : "")))
-        .build();
+    return URI.create(
+        SystemBaseUrl.EXTERNAL
+            .constructUrl("/logout?noPrompt=true")
+            .concat(requestQueryString != null ? "&" + requestQueryString : ""));
+  }
+
+  private long getTimeLeft(SecurityTokenHolder securityToken) {
+    List<SecurityAssertionImpl> values =
+        securityToken
+            .getRealmTokenMap()
+            .values()
+            .stream()
+            .map(SecurityAssertionImpl::new)
+            .collect(Collectors.toList());
+    values.sort(
```

Comment:
✏️ This whole, inefficient method builds a collection then sorts a collection then checks if the collection is empty; if not, it pulls the first, smallest, value out and processes against it. It could be simplified as something like...
```java
    return securityToken.getRealmTokenMap()
        .values()
        .stream()
        .map(SecurityAssertionImpl::new)
        .map(SecurityAssertionImpl::getNotOnOrAfter)
        .map(Date::getTime)
        .min(Comparator.comparing(Long::valueOf))
        .map(m -> Math.max(m - clock.millis(), 0))
        .orElse(0L);
```

(Should double check to confirm, of course; ideally by making this method package-private and adding unit tests for it then changing to this updated implementation.)